component/otelcol/exporter/exporter_test.go

```diff
@@ -0,0 +1,170 @@
+	Controller *componenttest.Controller
+}
+
+func newTestEnvironment(t *testing.T, fe *fakeExporter) *testEnvironment {
+	t.Helper()
+
+	reg := component.Registration{
+		Name:    "testcomponent",
+		Args:    fakeExporterArgs{},
+		Exports: otelcol.ConsumerExports{},
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			// Create a factory which always returns our instance of fakeExporter
+			// defined above.
+			factory := otelcomponent.NewExporterFactory(
+				"testcomponent",
+				func() otelconfig.Exporter {
+					return fakeExporterArgs{}.Convert()
+				},
+				otelcomponent.WithTracesExporter(func(ctx context.Context, ecs otelcomponent.ExporterCreateSettings, e otelconfig.Exporter) (otelcomponent.TracesExporter, error) {
+					return fe, nil
+				}),
+			)
+
+			return exporter.New(opts, factory, args.(exporter.Arguments))
+		},
+	}
+
+	return &testEnvironment{
+		t:          t,
+		Controller: componenttest.NewControllerFromReg(util.TestLogger(t), reg),
+	}
+}
+
+func (te *testEnvironment) Start() {
+	go func() {
+		ctx := componenttest.TestContext(te.t)
+		err := te.Controller.Run(ctx, fakeExporterArgs{})
+		require.NoError(te.t, err, "failed to run component")
+	}()
+}
+
+type fakeExporterArgs struct {
+}
+
+var _ exporter.Arguments = fakeExporterArgs{}
+
+func (fa fakeExporterArgs) Convert() otelconfig.Exporter {
+	settings := otelconfig.NewExporterSettings(otelconfig.NewComponentID("testcomponent"))
+	return &settings
+}
+
+func (fa fakeExporterArgs) Extensions() map[otelconfig.ComponentID]otelcomponent.Extension {
+	return nil
+}
+
+func (fa fakeExporterArgs) Exporters() map[otelconfig.DataType]map[otelconfig.ComponentID]otelcomponent.Exporter {
+	return nil
+}
+
+type fakeExporter struct {
+	StartFunc         func(ctx context.Context, host otelcomponent.Host) error
+	ShutdownFunc      func(ctx context.Context) error
+	CapabilitiesFunc  func() otelconsumer.Capabilities
+	ConsumeTracesFunc func(ctx context.Context, td ptrace.Traces) error
+}
+
+var _ otelcomponent.TracesExporter = (*fakeExporter)(nil)
+
+func (fe *fakeExporter) Start(ctx context.Context, host otelcomponent.Host) error {
+	if fe.StartFunc != nil {
+		return fe.StartFunc(ctx, host)
+	}
+	return nil
+}
+
+func (fe *fakeExporter) Shutdown(ctx context.Context) error {
+	if fe.ShutdownFunc != nil {
+		return fe.ShutdownFunc(ctx)
+	}
+	return nil
+}
+
+func (fe *fakeExporter) Capabilities() otelconsumer.Capabilities {
+	if fe.CapabilitiesFunc != nil {
+		return fe.CapabilitiesFunc()
+	}
+	return otelconsumer.Capabilities{}
+}
+
+func (fe *fakeExporter) ConsumeTraces(ctx context.Context, td ptrace.Traces) error {
+	if fe.ConsumeTracesFunc != nil {
+		return fe.ConsumeTracesFunc(ctx, td)
+	}
+	return nil
+}
+
+func createTestTraces() ptrace.Traces {
+	data := ptrace.NewTraces()
+	rss := data.ResourceSpans().AppendEmpty()
+	ss := rss.ScopeSpans().AppendEmpty()
+	s := ss.Spans().AppendEmpty()
+	s.SetName("TestSpan")
+
+	return data
+}
```

Comment:
This feels a bit silly, I'll look around for a JSON decoder to maybe make creating test traces a little easier.

Reply:
There's `ptrace.NewJSONUnmarshaler()`, which might be what I'm looking for here (as long as I can figure out what the JSON format looks like)